services/remotecluster/send.go

```diff
@@ -14,95 +14,120 @@ import (
+func (rcs *Service) BroadcastMsg(ctx context.Context, msg model.RemoteClusterMsg, f SendResultFunc) error {
+	// get list of interested remotes.
+	filter := model.RemoteClusterQueryFilter{
+		Topic: msg.Topic,
+	}
+	list, err := rcs.server.GetStore().RemoteCluster().GetAll(filter)
+	if err != nil {
+		return err
+	}
+
+	errs := merror.New()
+
+	for _, rc := range list {
+		if err := rcs.SendMsg(ctx, msg, rc, f); err != nil {
+			errs.Append(err)
+		}
 	}
+	return errs.ErrorOrNil()
+}
 
+// SendMsg asynchronously sends a message to a remote cluster.
+//
+// `ctx` determines behaviour when the outbound queue is full. A timeout or deadline context will return a
+// BufferFullError if the message cannot be enqueued before the timeout. A background context will block indefinitely.
+//
+// An optional callback can be provided that receives the success or fail result of sending to the remote cluster.
+// Success or fail is regarding message delivery only.  If a callback is provided it should return quickly.
+func (rcs *Service) SendMsg(ctx context.Context, msg model.RemoteClusterMsg, rc *model.RemoteCluster, f SendResultFunc) error {
 	if ctx == nil {
 		ctx = context.Background()
 	}
 
+	task := sendTask{
+		rc:  rc,
+		msg: msg,
+		f:   f,
+	}
+
 	select {
 	case rcs.send <- task:
 		return nil
 	case <-ctx.Done():
-		return fmt.Errorf("Could not enqueue outbound remote cluster message; queue full or timeout. Queue %d/%d. Err: %w", len(rcs.send), cap(rcs.send), ctx.Err())
+		return NewBufferFullError(cap(rcs.send))
 	}
+
 }
 
 func (rcs *Service) sendLoop(done chan struct{}) {
-	// create thread pool for concurrent message sending.
-	for i := 0; i < MaxConcurrentSends; i++ {
-		go func() {
-			for {
-				select {
-				case task := <-rcs.send:
-					rcs.sendMsg(task)
-				case <-done:
-					return
-				}
-			}
-		}()
+	for {
+		select {
+		case task := <-rcs.send:
+			rcs.sendMsg(task)
+		case <-done:
+			return
+		}
 	}
 }
 
 func (rcs *Service) sendMsg(task sendTask) {
-	// get list of interested remotes.
-	list, err := rcs.server.GetStore().RemoteCluster().GetByTopic(task.msg.Topic)
-	if err != nil && task.f != nil {
-		task.f(task.msg, nil, nil, fmt.Errorf("could not send msg to remote: %w", err))
-		return
+	// Ensure a panic from the callback doesn't exit the pool thread.
+	defer func() {
+		if r := recover(); r != nil {
+			rcs.server.GetLogger().Log(mlog.LvlPanic, "Remote Cluster sendMsg panic",
+				mlog.String("remote", task.rc.DisplayName), mlog.String("msgId", task.msg.Id), mlog.Any("panic", r))
+		}
+	}()
+
+	frame := &model.RemoteClusterFrame{
+		RemoteId: task.rc.RemoteId,
+		Token:    task.rc.RemoteToken,
+		Msg:      task.msg,
 	}
+	url := fmt.Sprintf("%s/%s", task.rc.SiteURL, SendMsgURL)
 
-	// bound the number of concurrent goroutines used to send using a semaphore.
-	bound := make(chan struct{}, MaxConcurrentSends)
+	resp, err := rcs.sendFrameToRemote(SendTimeout, frame, url)
+	response := make(Response)
```

Comment:
To me, it looks like we only need to create a response for the cases that a callback is provided. So in order to reduce some object allocations, we can assign the error string into a variable and only in the `if task.f != nil` condition to create the response object.

Reply:
The Response struct is also needed to unmarshall the response from the remote side regardless of callaback.